// ldkit/script/namespaces.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.xsd = exports.skos = exports.sioc = exports.schema = exports.rdfs = exports.rdf = exports.owl = exports.ldkit = exports.gr = exports.foaf = exports.dcterms = exports.dc = exports.dbo = void 0;
/**
 * Popular namespaces used in Linked Data, fully compatible with LDkit,
 * offering autocompletion and type checking in IDE.
 *
 * Create your own namespaces using {@link createNamespace} helper.
 *
 * @example
 * ```typescript
 * import { rdf, schema } from "ldkit/namespaces";
 *
 * console.log(rdf.type); // "http://www.w3.org/1999/02/22-rdf-syntax-ns#type"
 * console.log(schema.Person); // "http://schema.org/Person"
 * ```
 *
 * @module
 */
var dbo_js_1 = require("./namespaces/dbo.js");
Object.defineProperty(exports, "dbo", { enumerable: true, get: function () { return dbo_js_1.dbo; } });
var dc_js_1 = require("./namespaces/dc.js");
Object.defineProperty(exports, "dc", { enumerable: true, get: function () { return dc_js_1.dc; } });
var dcterms_js_1 = require("./namespaces/dcterms.js");
Object.defineProperty(exports, "dcterms", { enumerable: true, get: function () { return dcterms_js_1.dcterms; } });
var foaf_js_1 = require("./namespaces/foaf.js");
Object.defineProperty(exports, "foaf", { enumerable: true, get: function () { return foaf_js_1.foaf; } });
var gr_js_1 = require("./namespaces/gr.js");
Object.defineProperty(exports, "gr", { enumerable: true, get: function () { return gr_js_1.gr; } });
var ldkit_js_1 = require("./namespaces/ldkit.js");
Object.defineProperty(exports, "ldkit", { enumerable: true, get: function () { return ldkit_js_1.ldkit; } });
var owl_js_1 = require("./namespaces/owl.js");
Object.defineProperty(exports, "owl", { enumerable: true, get: function () { return owl_js_1.owl; } });
var rdf_js_1 = require("./namespaces/rdf.js");
Object.defineProperty(exports, "rdf", { enumerable: true, get: function () { return rdf_js_1.rdf; } });
var rdfs_js_1 = require("./namespaces/rdfs.js");
Object.defineProperty(exports, "rdfs", { enumerable: true, get: function () { return rdfs_js_1.rdfs; } });
var schema_js_1 = require("./namespaces/schema.js");
Object.defineProperty(exports, "schema", { enumerable: true, get: function () { return schema_js_1.schema; } });
var sioc_js_1 = require("./namespaces/sioc.js");
Object.defineProperty(exports, "sioc", { enumerable: true, get: function () { return sioc_js_1.sioc; } });
var skos_js_1 = require("./namespaces/skos.js");
Object.defineProperty(exports, "skos", { enumerable: true, get: function () { return skos_js_1.skos; } });
var xsd_js_1 = require("./namespaces/xsd.js");
Object.defineProperty(exports, "xsd", { enumerable: true, get: function () { return xsd_js_1.xsd; } });
__exportStar(require("./library/namespace.js"), exports);
