// ldkit/script/library/engine/query_engine_proxy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryEngineProxy = void 0;
const rdf_js_1 = require("../rdf.js");
class QueryEngineProxy {
    constructor(engine, context) {
        Object.defineProperty(this, "engine", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "context", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.engine = engine;
        this.context = context;
    }
    queryBoolean(query) {
        return this.engine.queryBoolean(query, this.context);
    }
    async queryBindings(query) {
        const bindingsStream = await this.engine.queryBindings(query, this.context);
        return bindingsStream.toArray();
    }
    async queryGraph(query) {
        const quadStream = await this.engine.queryQuads(query, this.context);
        return (0, rdf_js_1.quadsToGraph)(quadStream);
    }
    queryVoid(query) {
        return this.engine.queryVoid(query, this.context);
    }
}
exports.QueryEngineProxy = QueryEngineProxy;
