// ldkit/script/library/lens/update_helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateHelper = void 0;
const mod_js_1 = require("../sparql/mod.js");
const mod_js_2 = require("../schema/mod.js");
const encoder_js_1 = require("../encoder.js");
class UpdateHelper {
    constructor(schema, options, variableInitCounter = 0) {
        Object.defineProperty(this, "schema", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "properties", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "options", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "variableCounter", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
        Object.defineProperty(this, "deleteQuads", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "insertQuads", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "whereQuads", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        this.schema = schema;
        this.properties = (0, mod_js_2.getSchemaProperties)(schema);
        this.options = options;
        this.variableCounter = variableInitCounter;
    }
    process(entity) {
        for (const name in this.properties) {
            this.processProperty(entity, name, this.properties[name]);
        }
    }
    processProperty(entity, propertyName, property) {
        this.variableCounter++;
        const value = entity[propertyName];
        if (value === undefined) {
            return;
        }
        if (property["@array"]) {
            this.processArrayProperty(entity, propertyName, value);
        }
        else {
            this.processSingleProperty(entity, propertyName, value, property);
        }
    }
    processSingleProperty(entity, propertyName, propertyValue, property) {
        const deletePattern = {
            $id: entity.$id,
            [propertyName]: null,
        };
        const quadsToDelete = this.encode(deletePattern);
        this.deleteQuads.push(...quadsToDelete);
        if (property["@optional"]) {
            this.whereQuads.push((0, mod_js_1.OPTIONAL) `${quadsToDelete}`);
        }
        else {
            this.whereQuads.push(...quadsToDelete);
        }
        if (property["@optional"] && propertyValue === null) {
            // The intention was to delete a value of an optional property, nothing to insert
            return;
        }
        const insertPattern = {
            $id: entity.$id,
            [propertyName]: propertyValue,
        };
        const quadsToInsert = this.encode(insertPattern);
        this.insertQuads.push(...quadsToInsert);
    }
    processArrayProperty(entity, propertyName, propertyValue) {
        const config = this.parseArrayUpdateConfig(propertyValue);
        if (config.$set) {
            this.processArraySet(entity, propertyName, config.$set);
        }
        else {
            this.processArrayAddRemove(entity, propertyName, config.$add, config.$remove);
        }
    }
    processArraySet(entity, propertyName, propertyValue) {
        const deletePattern = {
            $id: entity.$id,
            [propertyName]: null,
        };
        const quadsToDelete = this.encode(deletePattern);
        this.deleteQuads.push(...quadsToDelete);
        this.whereQuads.push(...quadsToDelete);
        const insertPattern = {
            $id: entity.$id,
            [propertyName]: propertyValue,
        };
        const quadsToInsert = this.encode(insertPattern);
        this.insertQuads.push(...quadsToInsert);
    }
    processArrayAddRemove(entity, propertyName, $add, $remove) {
        if ($remove) {
            const deletePattern = {
                $id: entity.$id,
                [propertyName]: $remove,
            };
            const quadsToDelete = this.encode(deletePattern);
            this.deleteQuads.push(...quadsToDelete);
        }
        if ($add) {
            const insertPattern = {
                $id: entity.$id,
                [propertyName]: $add,
            };
            const quadsToInsert = this.encode(insertPattern);
            this.insertQuads.push(...quadsToInsert);
        }
    }
    parseArrayUpdateConfig(config) {
        if (Array.isArray(config)) {
            return { $set: config, $add: undefined, $remove: undefined };
        }
        if (config == null || typeof config !== "object") {
            throw new Error("Invalid array update query, expected an array, or an object with $add, $set, or $remove properties");
        }
        const { $add, $set, $remove } = config;
        switch (true) {
            case Array.isArray($set) && $add === undefined && $remove === undefined:
            case $set === undefined && Array.isArray($add) && $remove === undefined:
            case $set === undefined && $add === undefined && Array.isArray($remove):
            case $set === undefined && Array.isArray($add) && Array.isArray($remove):
                return { $add, $set, $remove };
        }
        throw new Error("Invalid array update query, expected an array, or an object with $add, $set, or $remove properties");
    }
    encode(entity) {
        return (0, encoder_js_1.encode)(entity, this.schema, this.options, false, this.variableCounter);
    }
}
exports.UpdateHelper = UpdateHelper;
