// ldkit/esm/library/engine/query_engine_proxy.js
import { quadsToGraph } from "../rdf.js";
export class QueryEngineProxy {
    constructor(engine, context) {
        Object.defineProperty(this, "engine", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "context", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.engine = engine;
        this.context = context;
    }
    queryBoolean(query) {
        return this.engine.queryBoolean(query, this.context);
    }
    async queryBindings(query) {
        const bindingsStream = await this.engine.queryBindings(query, this.context);
        return bindingsStream.toArray();
    }
    async queryGraph(query) {
        const quadStream = await this.engine.queryQuads(query, this.context);
        return quadsToGraph(quadStream);
    }
    queryVoid(query) {
        return this.engine.queryVoid(query, this.context);
    }
}
