// ldkit/esm/namespaces.js
/**
 * Popular namespaces used in Linked Data, fully compatible with LDkit,
 * offering autocompletion and type checking in IDE.
 *
 * Create your own namespaces using {@link createNamespace} helper.
 *
 * @example
 * ```typescript
 * import { rdf, schema } from "ldkit/namespaces";
 *
 * console.log(rdf.type); // "http://www.w3.org/1999/02/22-rdf-syntax-ns#type"
 * console.log(schema.Person); // "http://schema.org/Person"
 * ```
 *
 * @module
 */
export { dbo } from "./namespaces/dbo.js";
export { dc } from "./namespaces/dc.js";
export { dcterms } from "./namespaces/dcterms.js";
export { foaf } from "./namespaces/foaf.js";
export { gr } from "./namespaces/gr.js";
export { ldkit } from "./namespaces/ldkit.js";
export { owl } from "./namespaces/owl.js";
export { rdf } from "./namespaces/rdf.js";
export { rdfs } from "./namespaces/rdfs.js";
export { schema } from "./namespaces/schema.js";
export { sioc } from "./namespaces/sioc.js";
export { skos } from "./namespaces/skos.js";
export { xsd } from "./namespaces/xsd.js";
export * from "./library/namespace.js";
