// ldkit/script/library/sparql/sparql_update_builders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WITH = exports.DELETE = exports.INSERT = void 0;
const sparql_shared_builders_js_1 = require("./sparql_shared_builders.js");
class SparqlUpdateBuilder extends sparql_shared_builders_js_1.SparqlBuilder {
    WHERE(strings, ...values) {
        return this.template(strings, values, "WHERE", sparql_shared_builders_js_1.braces);
    }
    USING_NAMED(stringOrNamedNode) {
        return this.namedNode(stringOrNamedNode, "USING NAMED");
    }
    USING(stringOrNamedNode) {
        return this.namedNode(stringOrNamedNode, "USING");
    }
    INSERT(strings, ...values) {
        return this.template(strings, values, "INSERT", sparql_shared_builders_js_1.braces);
    }
    INSERT_DATA(strings, ...values) {
        return this.template(strings, values, "INSERT DATA", sparql_shared_builders_js_1.braces);
    }
    DELETE(strings, ...values) {
        return this.template(strings, values, "DELETE", sparql_shared_builders_js_1.braces);
    }
    DELETE_DATA(strings, ...values) {
        return this.template(strings, values, "DELETE DATA", sparql_shared_builders_js_1.braces);
    }
    DELETE_WHERE(strings, ...values) {
        return this.template(strings, values, "DELETE WHERE", sparql_shared_builders_js_1.braces);
    }
    WITH(stringOrNamedNode) {
        return this.namedNode(stringOrNamedNode, "WITH");
    }
}
/**
 * SPARQL INSERT query fluent interface
 *
 * @example
 * ```typescript
 * import { INSERT } from "ldkit/sparql";
 * import { foaf } from "ldkit/namespaces";
 * import { DataFactory } from "ldkit/rdf";
 *
 * const df = new DataFactory();
 * const firstName = df.namedNode(foaf.firstName);
 *
 * const query = INSERT`?person ${firstName} "Paul"`
 *   .WHERE`?person ${firstName} "Jean"`
 *   .build();
 * console.log(query);
 * // INSERT { ?person <http://xmlns.com/foaf/0.1/firstName> "Paul" }
 * // WHERE { ?person <http://xmlns.com/foaf/0.1/firstName> "Jean" }
 * ```
 */
exports.INSERT = Object.assign((strings, ...values) => new SparqlUpdateBuilder().INSERT(strings, ...values), {
    DATA: (strings, ...values) => new SparqlUpdateBuilder().INSERT_DATA(strings, ...values),
});
/**
 * SPARQL DELETE query fluent interface
 *
 * @example
 * ```typescript
 * import { DELETE } from "ldkit/sparql";
 * import { foaf } from "ldkit/namespaces";
 * import { DataFactory } from "ldkit/rdf";
 *
 * const df = new DataFactory();
 * const firstName = df.namedNode(foaf.firstName);
 *
 * const query = DELETE`?person ${firstName} "Jean"`
 *   .INSERT`?person ${firstName} "Paul"`
 *   .WHERE`?person ${firstName} "Jean"`
 *   .build();
 * console.log(query);
 * // DELETE { ?person <http://xmlns.com/foaf/0.1/firstName> "Jean" }
 * // INSERT { ?person <http://xmlns.com/foaf/0.1/firstName> "Paul" }
 * // WHERE { ?person <http://xmlns.com/foaf/0.1/firstName> "Jean" }
 * ```
 */
exports.DELETE = Object.assign((strings, ...values) => new SparqlUpdateBuilder().DELETE(strings, ...values), {
    DATA: (strings, ...values) => new SparqlUpdateBuilder().DELETE_DATA(strings, ...values),
    WHERE: (strings, ...values) => new SparqlUpdateBuilder().DELETE_WHERE(strings, ...values),
});
/**
 * SPARQL WITH query fluent interface
 *
 * @example
 * ```typescript
 * import { DELETE } from "ldkit/sparql";
 * import { foaf } from "ldkit/namespaces";
 * import { DataFactory } from "ldkit/rdf";
 *
 * const df = new DataFactory();
 * const firstName = df.namedNode(foaf.firstName);
 * const graph = df.namedNode("http://example.org/graph");
 *
 * const query = WITH(graph).DELETE`?person ${firstName} "Jean"`
 *   .INSERT`?person ${firstName} "Paul"`
 *   .WHERE`?person ${firstName} "Jean"`
 *   .build();
 * console.log(query);
 * // WITH <http://example.org/graph>
 * // DELETE { ?person <http://xmlns.com/foaf/0.1/firstName> "Jean" }
 * // INSERT { ?person <http://xmlns.com/foaf/0.1/firstName> "Paul" }
 * // WHERE { ?person <http://xmlns.com/foaf/0.1/firstName> "Jean" }
 * ```
 */
const WITH = (stringOrNamedNode) => new SparqlUpdateBuilder().WITH(stringOrNamedNode);
exports.WITH = WITH;
