// ldkit/script/library/sparql/sparql_shared_builders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SparqlBuilder = exports.parentheses = exports.bracesInline = exports.braces = void 0;
const rdf_js_1 = require("../rdf.js");
const sparql_tag_js_1 = require("./sparql_tag.js");
const braces = (keyword, value) => `${keyword} {\n${value}\n}\n`;
exports.braces = braces;
const bracesInline = (keyword, value) => `${keyword} { ${value} }`;
exports.bracesInline = bracesInline;
const parentheses = (keyword, value) => `${keyword} (${value})\n`;
exports.parentheses = parentheses;
const none = (keyword, value) => `${keyword} ${value}\n`;
class SparqlBuilder {
    constructor() {
        Object.defineProperty(this, "value", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: ""
        });
        Object.defineProperty(this, "dataFactory", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new rdf_js_1.DataFactory()
        });
    }
    build() {
        return this.value;
    }
    sparql(strings, ...values) {
        return !values || values.length < 1
            ? (0, sparql_tag_js_1.sparql)(strings)
            : (0, sparql_tag_js_1.sparql)(strings, ...values);
    }
    template(strings, values, keyword, wrap = none) {
        const inputSparql = this.sparql(strings, ...values);
        this.value += wrap(keyword, inputSparql);
        return this;
    }
    namedNode(stringOrNamedNode, keyword, wrap = none) {
        const namedNode = typeof stringOrNamedNode === "string"
            ? this.dataFactory.namedNode(stringOrNamedNode)
            : stringOrNamedNode;
        const inputSparql = this.sparql `${namedNode}`;
        this.value += wrap(keyword, inputSparql);
        return this;
    }
    number(number, keyword, wrap = none) {
        this.value += wrap(keyword, number.toString());
        return this;
    }
}
exports.SparqlBuilder = SparqlBuilder;
