// ldkit/script/library/sparql/sparql_query_builders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DESCRIBE = exports.ASK = exports.CONSTRUCT = exports.SELECT = void 0;
const sparql_shared_builders_js_1 = require("./sparql_shared_builders.js");
class SparqlQueryBuilder extends sparql_shared_builders_js_1.SparqlBuilder {
    OFFSET(number) {
        return this.number(number, "OFFSET");
    }
    LIMIT(number) {
        return this.number(number, "LIMIT");
    }
    ORDER_BY(strings, ...values) {
        return this.template(strings, values, "ORDER BY");
    }
    HAVING(strings, ...values) {
        return this.template(strings, values, "HAVING", sparql_shared_builders_js_1.parentheses);
    }
    GROUP_BY(strings, ...values) {
        return this.template(strings, values, "GROUP BY");
    }
    WHERE(strings, ...values) {
        return this.template(strings, values, "WHERE", sparql_shared_builders_js_1.braces);
    }
    FROM_NAMED(stringOrNamedNode) {
        return this.namedNode(stringOrNamedNode, "FROM NAMED");
    }
    FROM(stringOrNamedNode) {
        return this.namedNode(stringOrNamedNode, "FROM");
    }
    SELECT(strings, ...values) {
        return this.template(strings, values, "SELECT");
    }
    SELECT_DISTINCT(strings, ...values) {
        return this.template(strings, values, "SELECT DISTINCT");
    }
    SELECT_REDUCED(strings, ...values) {
        return this.template(strings, values, "SELECT REDUCED");
    }
    CONSTRUCT(strings, ...values) {
        return this.template(strings, values, "CONSTRUCT", sparql_shared_builders_js_1.braces);
    }
    CONSTRUCT_WHERE(strings, ...values) {
        return this.template(strings, values, "CONSTRUCT WHERE", sparql_shared_builders_js_1.braces);
    }
    ASK(strings, ...values) {
        return this.template(strings, values, "ASK", sparql_shared_builders_js_1.braces);
    }
    ASK_FROM(stringOrNamedNode) {
        return this.namedNode(stringOrNamedNode, "ASK\nFROM");
    }
    ASK_FROM_NAMED(stringOrNamedNode) {
        return this.namedNode(stringOrNamedNode, "ASK\nFROM NAMED");
    }
    ASK_WHERE(strings, ...values) {
        return this.template(strings, values, "ASK\nWHERE", sparql_shared_builders_js_1.braces);
    }
    DESCRIBE(strings, ...values) {
        return this.template(strings, values, "DESCRIBE");
    }
}
/**
 * SPARQL SELECT query fluent interface
 *
 * @example
 * ```typescript
 * import { SELECT } from "ldkit/sparql";
 *
 * const query = SELECT`?s`.WHERE`?s ?p ?o`.ORDER_BY`?s`.LIMIT(100).build();
 * console.log(query);
 * // SELECT ?s WHERE { ?s ?p ?o } ORDER BY ?s LIMIT 100
 * ```
 *
 * @example
 * ```typescript
 * import { SELECT } from "ldkit/sparql";
 *
 * const query = SELECT.DISTINCT`?s`.WHERE`?s ?p ?o`.build();
 * console.log(query);
 * // SELECT DISTINCT ?s WHERE { ?s ?p ?o }
 * ```
 */
exports.SELECT = Object.assign((strings, ...values) => new SparqlQueryBuilder().SELECT(strings, ...values), {
    DISTINCT: (strings, ...values) => new SparqlQueryBuilder().SELECT_DISTINCT(strings, ...values),
    REDUCED: (strings, ...values) => new SparqlQueryBuilder().SELECT_REDUCED(strings, ...values),
    get ALL() {
        return new SparqlQueryBuilder().SELECT `*`;
    },
});
/**
 * SPARQL CONSTRUCT query fluent interface
 *
 * @example
 * ```typescript
 * import { CONSTRUCT } from "ldkit/sparql";
 * import { DataFactory } from "ldkit/rdf";
 *
 * const df = new DataFactory();
 * const sNode = df.namedNode("http://example.org/datasource");
 * const pNode = df.namedNode("http://example.org/hasSubject");
 *
 * const query = CONSTRUCT`${sNode} ${pNode} ?s`.WHERE`?s ?p ?o`.build();
 * console.log(query);
 * // CONSTRUCT { <http://example.org/datasource> <http://example.org/hasSubject> ?s }
 * // WHERE { ?s ?p ?o }
 * ```
 */
exports.CONSTRUCT = Object.assign((strings, ...values) => new SparqlQueryBuilder().CONSTRUCT(strings, ...values), {
    WHERE: (strings, ...values) => new SparqlQueryBuilder().CONSTRUCT_WHERE(strings, ...values),
});
/**
 * SPARQL ASK query fluent interface
 *
 * @example
 * ```typescript
 * import { ASK } from "ldkit/sparql";
 *
 * const query = ASK`?s ?p ?o`.build();
 * console.log(query); // ASK { ?s ?p ?o }
 * ```
 */
exports.ASK = Object.assign((strings, ...values) => new SparqlQueryBuilder().ASK(strings, ...values), {
    FROM: (stringOrNamedNode) => new SparqlQueryBuilder().ASK_FROM(stringOrNamedNode),
    FROM_NAMED: (stringOrNamedNode) => new SparqlQueryBuilder().ASK_FROM_NAMED(stringOrNamedNode),
    WHERE: (strings, ...values) => new SparqlQueryBuilder().ASK_WHERE(strings, ...values),
});
/**
 * SPARQL DESCRIBE query fluent interface
 *
 * @example
 * ```typescript
 * import { DESCRIBE } from "ldkit/sparql";
 * import { DataFactory } from "ldkit/rdf";
 *
 * const df = new DataFactory();
 * const node = df.namedNode("http://example.org/resource");
 *
 * const query = DESCRIBE`${node}`.build();
 * console.log(query); // DESCRIBE <http://example.org/resource>
 * ```
 */
const DESCRIBE = (strings, ...values) => new SparqlQueryBuilder().DESCRIBE(strings, ...values);
exports.DESCRIBE = DESCRIBE;
