// ldkit/script/library/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setGlobalOptions = setGlobalOptions;
exports.resolveOptions = resolveOptions;
exports.resolveQueryContext = resolveQueryContext;
const mod_js_1 = require("./engine/mod.js");
const defaultOptions = {
    engine: new mod_js_1.QueryEngine(),
    take: 1000,
    logQuery: () => { },
};
let globalOptions = {};
/**
 * Sets global configuration {@link Options} for LDkit that will be used
 * by default in all queries, unless overridden in {@link Lens}.
 *
 * LDkit-specific options are:
 * - `engine` - a query engine to use for querying data sources
 * - `language` - a preferred language for literals
 * - `take` - a default number of results to take (limit of SELECT queries)
 * - `logQuery` - a function that will be called for each SPARQL query
 *
 * Default values for these options are:
 * ```typescript
 * const defaultOptions = {
 *   engine: new QueryEngine(),
 *   take: 1000,
 *   logQuery: () => {},
 * };
 * ```
 * The default configuration uses built-in {@link QueryEngine}. Language is not set by default.
 *
 * @param options LDkit options and query engine context
 */
function setGlobalOptions(options) {
    globalOptions = options;
}
function resolveOptions(options = {}) {
    return {
        ...defaultOptions,
        ...globalOptions,
        ...options,
    };
}
function resolveQueryContext(options) {
    const { engine: _engine, language: _language, take: _take, ...context } = options;
    if (context.source !== undefined && context.sources === undefined) {
        context.sources = [context.source];
        delete context.source;
    }
    if (context.sources === undefined) {
        throw new Error("No data source defined in Options");
    }
    return context;
}
