// ldkit/esm/library/sparql/sparql_shared_builders.js
import { DataFactory } from "../rdf.js";
import { sparql } from "./sparql_tag.js";
export const braces = (keyword, value) => `${keyword} {\n${value}\n}\n`;
export const bracesInline = (keyword, value) => `${keyword} { ${value} }`;
export const parentheses = (keyword, value) => `${keyword} (${value})\n`;
const none = (keyword, value) => `${keyword} ${value}\n`;
export class SparqlBuilder {
    constructor() {
        Object.defineProperty(this, "value", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: ""
        });
        Object.defineProperty(this, "dataFactory", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new DataFactory()
        });
    }
    build() {
        return this.value;
    }
    sparql(strings, ...values) {
        return !values || values.length < 1
            ? sparql(strings)
            : sparql(strings, ...values);
    }
    template(strings, values, keyword, wrap = none) {
        const inputSparql = this.sparql(strings, ...values);
        this.value += wrap(keyword, inputSparql);
        return this;
    }
    namedNode(stringOrNamedNode, keyword, wrap = none) {
        const namedNode = typeof stringOrNamedNode === "string"
            ? this.dataFactory.namedNode(stringOrNamedNode)
            : stringOrNamedNode;
        const inputSparql = this.sparql `${namedNode}`;
        this.value += wrap(keyword, inputSparql);
        return this;
    }
    number(number, keyword, wrap = none) {
        this.value += wrap(keyword, number.toString());
        return this;
    }
}
