// ldkit/esm/library/sparql/sparql_update_builders.js
import { braces, SparqlBuilder } from "./sparql_shared_builders.js";
class SparqlUpdateBuilder extends SparqlBuilder {
    WHERE(strings, ...values) {
        return this.template(strings, values, "WHERE", braces);
    }
    USING_NAMED(stringOrNamedNode) {
        return this.namedNode(stringOrNamedNode, "USING NAMED");
    }
    USING(stringOrNamedNode) {
        return this.namedNode(stringOrNamedNode, "USING");
    }
    INSERT(strings, ...values) {
        return this.template(strings, values, "INSERT", braces);
    }
    INSERT_DATA(strings, ...values) {
        return this.template(strings, values, "INSERT DATA", braces);
    }
    DELETE(strings, ...values) {
        return this.template(strings, values, "DELETE", braces);
    }
    DELETE_DATA(strings, ...values) {
        return this.template(strings, values, "DELETE DATA", braces);
    }
    DELETE_WHERE(strings, ...values) {
        return this.template(strings, values, "DELETE WHERE", braces);
    }
    WITH(stringOrNamedNode) {
        return this.namedNode(stringOrNamedNode, "WITH");
    }
}
/**
 * SPARQL INSERT query fluent interface
 *
 * @example
 * ```typescript
 * import { INSERT } from "ldkit/sparql";
 * import { foaf } from "ldkit/namespaces";
 * import { DataFactory } from "ldkit/rdf";
 *
 * const df = new DataFactory();
 * const firstName = df.namedNode(foaf.firstName);
 *
 * const query = INSERT`?person ${firstName} "Paul"`
 *   .WHERE`?person ${firstName} "Jean"`
 *   .build();
 * console.log(query);
 * // INSERT { ?person <http://xmlns.com/foaf/0.1/firstName> "Paul" }
 * // WHERE { ?person <http://xmlns.com/foaf/0.1/firstName> "Jean" }
 * ```
 */
export const INSERT = Object.assign((strings, ...values) => new SparqlUpdateBuilder().INSERT(strings, ...values), {
    DATA: (strings, ...values) => new SparqlUpdateBuilder().INSERT_DATA(strings, ...values),
});
/**
 * SPARQL DELETE query fluent interface
 *
 * @example
 * ```typescript
 * import { DELETE } from "ldkit/sparql";
 * import { foaf } from "ldkit/namespaces";
 * import { DataFactory } from "ldkit/rdf";
 *
 * const df = new DataFactory();
 * const firstName = df.namedNode(foaf.firstName);
 *
 * const query = DELETE`?person ${firstName} "Jean"`
 *   .INSERT`?person ${firstName} "Paul"`
 *   .WHERE`?person ${firstName} "Jean"`
 *   .build();
 * console.log(query);
 * // DELETE { ?person <http://xmlns.com/foaf/0.1/firstName> "Jean" }
 * // INSERT { ?person <http://xmlns.com/foaf/0.1/firstName> "Paul" }
 * // WHERE { ?person <http://xmlns.com/foaf/0.1/firstName> "Jean" }
 * ```
 */
export const DELETE = Object.assign((strings, ...values) => new SparqlUpdateBuilder().DELETE(strings, ...values), {
    DATA: (strings, ...values) => new SparqlUpdateBuilder().DELETE_DATA(strings, ...values),
    WHERE: (strings, ...values) => new SparqlUpdateBuilder().DELETE_WHERE(strings, ...values),
});
/**
 * SPARQL WITH query fluent interface
 *
 * @example
 * ```typescript
 * import { DELETE } from "ldkit/sparql";
 * import { foaf } from "ldkit/namespaces";
 * import { DataFactory } from "ldkit/rdf";
 *
 * const df = new DataFactory();
 * const firstName = df.namedNode(foaf.firstName);
 * const graph = df.namedNode("http://example.org/graph");
 *
 * const query = WITH(graph).DELETE`?person ${firstName} "Jean"`
 *   .INSERT`?person ${firstName} "Paul"`
 *   .WHERE`?person ${firstName} "Jean"`
 *   .build();
 * console.log(query);
 * // WITH <http://example.org/graph>
 * // DELETE { ?person <http://xmlns.com/foaf/0.1/firstName> "Jean" }
 * // INSERT { ?person <http://xmlns.com/foaf/0.1/firstName> "Paul" }
 * // WHERE { ?person <http://xmlns.com/foaf/0.1/firstName> "Jean" }
 * ```
 */
export const WITH = (stringOrNamedNode) => new SparqlUpdateBuilder().WITH(stringOrNamedNode);
